src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -0,0 +1,208 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace ray {
+namespace syncer {
+
+/// NodeState keeps track of the modules in the local nodes.
+/// It contains the local components for receiving and reporting.
+/// It also keeps the raw messages receivers got.
+class NodeState {
+ public:
+  /// Constructor of NodeState.
+  NodeState();
+
+  /// Set the local component.
+  ///
+  /// \param cid The component id.
+  /// \param reporter The reporter is defined to be the local module which wants to
+  /// broadcast its internal status to the whole clsuter. When it's null, it means there
+  /// is no reporter in this node for this component. This is the place there messages are
+  /// generated.
+  /// \param receiver The receiver is defined to be the module which eventually
+  /// will have the view of of the cluster for this component. It's the place where
+  /// received messages are consumed.
+  ///
+  /// \return true if set successfully.
+  bool SetComponent(RayComponentId cid,
+                    const ReporterInterface *reporter,
+                    ReceiverInterface *receiver);
+
+  /// Get the snapshot of a component for a newer version.
+  ///
+  /// \param cid The component id to take the snapshot.
+  ///
+  /// \return If a snapshot is taken, return the message, otherwise std::nullopt.
+  std::optional<RaySyncMessage> GetSnapshot(RayComponentId cid);
+
+  /// Consume a message. Receiver will consume this message if it doesn't have
+  /// this message.
+  ///
+  /// \param message The message received.
+  ///
+  /// \return true if this node doesn't have message with newer version.
+  bool ConsumeMessage(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Return the cluster view of this local node.
+  const absl::flat_hash_map<
+      std::string,
+      std::array<std::shared_ptr<const RaySyncMessage>, kComponentArraySize>>
+      &GetClusterView() const {
+    return cluster_view_;
+  }
+
+ private:
+  /// For local nodes
+  std::array<const ReporterInterface *, kComponentArraySize> reporters_ = {nullptr};
+  std::array<ReceiverInterface *, kComponentArraySize> receivers_ = {nullptr};
+
+  /// This field records the version of the snapshot that has been taken.
+  std::array<int64_t, kComponentArraySize> snapshots_versions_taken_;
+  /// Keep track of the latest messages received.
+  /// Use shared pointer for easier liveness management since these messages might be
+  /// sending via rpc.
+  absl::flat_hash_map<
+      std::string,
+      std::array<std::shared_ptr<const RaySyncMessage>, kComponentArraySize>>
+      cluster_view_;
+};
+
+class NodeSyncConnection {
```

Comment:
think again about this NodeSyncConnection. If we are using it to simulating streaming, should it have following APIs?
```
class NodeSyncConnection {
  // Receives stream of responses from remote node. 
  using ResponseObserver = std::function<void(Response)>;
  public:
   NodeSyncConnection(ResponseObserver response_observer);
   // Sends requests to remote node.
   bool SendRequest(Request request);
};
```

Reply:
@scv119  even with streaming, we still need Server and Client concept here. The one who initializes the work is client.

Here the difference is: 1) initialization; 2) how to send; 3) how to receive. All these are communication layer things.

Put them into NodeSyncConnection gives me the feeling that it's not a good abstraction.

Now it's like:
- NodeSyncConnection -> application layer filtering
- Server/ClientSyncConneciton-> Handling how to do communication.

Reply:
hmm could you hide them under the NodeSyncConnection interface/abstract class?
```
class NodeSyncConnection {
  // Receives stream of responses from remote node. 
  using ResponseObserver = std::function<void(Response)>;
  public:
   NodeSyncConnection(ResponseObserver response_observer);
   // Sends requests to remote node.
   bool SendRequest(Request request);
};

class ServerSyncConnection : public NodeSyncConnection {
 public:
   ServerSyncConnection(ResponseObserver response_observer);
   bool SendRequest(Request request) override;
};

class ClientSyncConnection : public NodeSyncConnection {
 public:
   ServerSyncConnection(ResponseObserver response_observer);
   bool SendRequest(Request request) override;
};

```

Reply:
I think it's not changing anything.

- PushToSendingQueue is SendRequest
- ReceiveUpdates is ResponseObserver

Btw, the only different part is about communication, what to do when we receive the updates is the same.

And we still end up having two subclasses of NodeSyncConnection.

Once, thing I think we probably can do is DoSend is a callback. But is this better than the current implementation?

Reply:
I just updated it. I think it looks better than before.